src/coord/src/coord.rs

```diff
@@ -5263,3 +5187,94 @@ impl Coordinator {
             self.finalize_dataflow(df, DEFAULT_COMPUTE_INSTANCE_ID);
     }
 }
+
+/// A mechanism to ensure that a sequence of writes and reads proceed correctly through timestamps.
+mod timeline {
+
+    /// A timeline is either currently writing or currently reading.
+    ///
+    /// At each time, writes happen and then reads happen, meaning that writes at a time are
+    /// visible to exactly reads at that time or greater, and no other times.
+    enum TimelineState {
+        /// The timeline is producing collection updates timestamped with the argument.
+        Writing(mz_repr::Timestamp),
+        /// The timeline is observing collections aot the time of the argument.
+        Reading(mz_repr::Timestamp),
+    }
+
+    /// A type that provides write and read timestamps, reads observe exactly their preceding writes..
+    ///
+    /// Specifically, all read timestamps will be greater or equal to all previously reported write timestamps,
+    /// and strictly less than all subsequently emitted write timestamps.
+    pub struct Timeline {
+        state: TimelineState,
+        advance_to: Option<mz_repr::Timestamp>,
+    }
+
+    impl Timeline {
+        /// Create a new timeline, starting at the indicated time.
+        pub fn new(initially: mz_repr::Timestamp) -> Self {
+            Self {
+                state: TimelineState::Writing(initially),
+                advance_to: Some(initially),
+            }
+        }
+
+        /// Acquire a new timestamp for writing.
+        ///
+        /// This timestamp will be strictly greater than all prior values of `self.read_ts()`,
+        /// and less than or equal to all subsequent values of `self.read_ts()`.
+        pub fn write_ts(&mut self) -> mz_repr::Timestamp {
+            match self.state {
+                TimelineState::Writing(ts) => ts,
+                TimelineState::Reading(ts) => {
+                    self.state = TimelineState::Writing(ts + 1);
```

Comment:
`ts+1` makes me sad because it interacts poorly with `fast_forward` (and other things where time is provided externally). #10173 is an old proposal to have `write_ts` functions return an `Option<Timestamp>`, so that they can signal to callers "I wasn't able to get a new write timestamp, try again later?". I think that would require having this struct grow a method or an enum telling it how to determine the next timestamp (by a func, just add 1).

Reply:
We could certainly change it (e.g. to have `None` mean: no timestamp for you; please call `fast_forward` with a greater time). In what way does +1 make you sad?

Reply:
I'm aware of two ways where the timestamp generated by this function will start producing timestamps in advance of now() due to `read_ts()` and `write_ts()` to forcibly incrementing by 1:

1. Repeated calls SELECT/INSERT over tables in a loop at a rate faster than once per millisecond.
2. If the system clock goes backward.

For 2, RT sources will bundle all data until now catches back up. But tables will continue to mint new timestamps.

I believe that for both of these cases the correct behavior is to tell callers "nope, you decide what to do". User things can block and wait, system things (like metrics) can be dropped on the floor (Andreas told me this is fine).

Reply:
An additional reason this is bad: `fast_forward` is only called once per second by the "make sure times go forward" loop. It's the only thing that cares about "now" for table timestamps. Otherwise they just get incremented by 1. That means that when users do writes to tables, the write timestamp is: "(now() from at most 1s ago) + (number of write-then-reads done since then)milliseconds".

For example, if the fast_forward loop happens to be called at exactly the second boundary, then writes will appear at 1.000 then 1.001 then 1.002 ... until the fast_forward loop is hit again, then 2.000, 2.001, etc. In this case write times will be behind now() unless they write faster than once per ms or the clock goes backward, at which point they will be ahead of now(). So they will almost never be at now().

Reply:
Here's logs of two sessions illustrating this:

```
materialize=> create table t (i int);
CREATE TABLE
materialize=> copy (tail t with (progress)) to stdout;
1646768248434   t       \N      \N
1646768248435   t       \N      \N
1646768248435   f       1       1
1646768248436   t       \N      \N
1646768248436   f       1       1
1646768248437   t       \N      \N
1646768248437   f       1       1
1646768248438   t       \N      \N
1646768248438   f       1       1
1646768248439   t       \N      \N
1646768248439   f       1       1
1646768248440   t       \N      \N
```

```
materialize=> insert into t values(1); select count(*) from t;
INSERT 0 1
 count 
-------
     1
(1 row)

materialize=> insert into t values(1); select count(*) from t;
INSERT 0 1
 count 
-------
     2
(1 row)

materialize=> insert into t values(1); select count(*) from t;
INSERT 0 1
 count 
-------
     3
(1 row)

materialize=> insert into t values(1); select count(*) from t;
INSERT 0 1
 count 
-------
     4
(1 row)

materialize=> insert into t values(1); select count(*) from t;
INSERT 0 1
 count 
-------
     5
(1 row)
```